saleor/warehouse/models.py

```diff
@@ -138,29 +178,67 @@ def applicable_for_click_and_collect(
             .select_related("product_variant")
         )
 
-        return self._for_channel_lines_and_stocks(lines_qs, stocks_qs, channel_id)
+        number_of_variants = (
+            lines_qs.order_by("variant_id").distinct("variant_id").count()
+        )
+        warehouses_for_channel = self.for_channel(channel_id)
+        warehouse_ids_with_stock_available = self._for_channel_lines_and_stocks(
+            number_of_variants, stocks_qs, channel_id
+        ).values("id")
+        # if there is any valid local warehouse it means that it is possible
+        # to ship products to any warehouse with `all warehouses` option
+        if warehouse_ids_with_stock_available.exists():
+            return warehouses_for_channel.filter(
+                Q(id__in=warehouse_ids_with_stock_available)
+                | Q(click_and_collect_option=warehouse_cc_option_enum.ALL_WAREHOUSES)
+            )
+
+        # Check if the ordered line quantities can be fulfilled using stock from
+        # different warehouses.
+        # First, stock is annotated with the `total_available_quantity` variable, which
+        # represents the total available quantity for each variant.
+        # Next, the queryset is annotated with `line_quantity`, indicating whether
+        # the available stock can cover the ordered line quantities.
+        # A positive `line_quantity` value means the order can be fulfilled.
+        stocks_qs = (
```

Comment:
Can you share in comment the sql query that we will get from this one?
It feels like a heavy one.

Reply:
You are right, query is quite complicated:
```
SELECT 
  "warehouse_stock"."product_variant_id", 
  SUM(
    (
      "warehouse_stock"."quantity" - COALESCE(
        (
          SELECT 
            SUM(U0."quantity_allocated") AS "total_allocated_quantity" 
          FROM 
            "warehouse_allocation" U0 
          WHERE 
            U0."stock_id" = ("warehouse_stock"."id") 
          GROUP BY 
            U0."stock_id"
        ), 
        0
      )
    )
  ) AS "total_available_quantity" 
FROM 
  "warehouse_stock" 
WHERE 
  "warehouse_stock"."product_variant_id" IN (
    SELECT 
      U0."variant_id" 
    FROM 
      "checkout_checkoutline" U0 
    WHERE 
      U0."checkout_id" = 1e718c01 - 4753 - 4942 - 8dfe - 306627be9512
  ) 
GROUP BY 
  "warehouse_stock"."product_variant_id" 
HAVING 
  (
    SUM(
      (
        "warehouse_stock"."quantity" - COALESCE(
          (
            SELECT 
              SUM(U0."quantity_allocated") AS "total_allocated_quantity" 
            FROM 
              "warehouse_allocation" U0 
            WHERE 
              U0."stock_id" = ("warehouse_stock"."id") 
            GROUP BY 
              U0."stock_id"
          ), 
          0
        )
      )
    ) - (
      SELECT 
        SUM(U0."quantity") AS "prod_sum" 
      FROM 
        "checkout_checkoutline" U0 
      WHERE 
        (
          U0."checkout_id" = 1e718c01 - 4753 - 4942 - 8dfe - 306627be9512 
          AND U0."variant_id" = (
            "warehouse_stock"."product_variant_id"
          )
        ) 
      GROUP BY 
        U0."variant_id" 
      ORDER BY 
        U0."variant_id" ASC
    )
  ) >= 0
```

Alternatively, I can annotate stocks with `annotate_total_available_quantity_per_variant` and then use for loop to check if all lines quantities are covered